app/views/play/level/tome/editor/autocomplete.coffee

```diff
@@ -314,6 +314,54 @@ module.exports = class Autocomplete
     haveFindNearest = false
     autocompleteReplacement = level.get("autocompleteReplacement") ? []
     usedAutocompleteReplacement = []
+
+    fixlanguageSnippets = (doc, lang) ->
+      usedAutocompleteReplacement = []
+
+      if lang in ['java', 'cpp'] and not doc?.snippets?[lang] and doc?.snippets?.javascript
+        doc.snippets[lang] = doc.snippets.javascript
+
+      if lang in ['lua', 'coffeescript', 'python'] and not doc?.snippets?[lang] and (doc?.snippets?.python or doc?.snippets?.javascript)
+          # These are mostly the same, so use the Python or JavaScript ones if language-specific ones aren't available
+        doc.snippets[lang] = doc?.snippets?.python or doc.snippets.javascript
+
+      if doc?.snippets?[lang]
+        name = doc.name
+        replacement = _.find(autocompleteReplacement, (el) -> el.name is name)
+        if replacement
+          usedAutocompleteReplacement.push(replacement.name)
+        content = replacement?.snippets?[lang]?.code or doc.snippets[lang].code
+        if /loop/.test(content) and level.get 'moveRightLoopSnippet'
+          # Replace default loop snippet with an embedded moveRight()
+          content = switch lang
+            when 'python' then 'while True:\n    hero.moveRight()\n    ${1:}'
+            when 'javascript', 'java', 'cpp' then 'while (true) {\n    hero.moveRight();\n    ${1:}\n}'
+            else content
+        if /loop/.test(content) and level.isType('course', 'course-ladder')
+          # Temporary hackery to make it look like we meant while True: in our loop snippets until we can update everything
+          content = switch lang
+            when 'python' then content.replace /loop:/, 'while True:'
+            when 'javascript', 'java', 'cpp' then content.replace /loop/, 'while (true)'
+            when 'lua' then content.replace /loop/, 'while true then'
+            when 'coffeescript' then content
+            else content
+          name = switch lang
+            when 'python' then 'while True'
+            when 'coffeescript' then 'loop'
+            else 'while true'
+        # For now, update autocomplete to use hero instead of self/this, if hero is already used in the source.
+        # Later, we should make this happen all the time - or better yet update the snippets.
+        if /hero/.test(source) or not /(self[\.\:]|this\.|\@)/.test(source)
+          thisToken =
+            'python': /self/,
+            'javascript': /this/,
+            'java': /this/,
+            'cpp': /this/,
+            'lua': /self/
+          if thisToken[lang] and thisToken[lang].test(content)
+            content = content.replace thisToken[lang], 'hero'
+       return [doc, content, name]
```

Comment:
This works, but would be better style to return as an object, rather than relying on the ordered aray.